schema.md

```diff
@@ -0,0 +1,136 @@
+# Sparkle Motion schema (WIP)
+
+All record types have `createdAt` (and maybe `modifiedAt`?).
+
+### User
+
+```js
+{
+  "id": ___,
+  "username": "xian",
+  "password": "...",      // salted hash of pw
+  "name": "Full Name",
+  "priority": 123.4,      // e.g. for locking out other users during a performance
+}
+```
+
+### Gadget Presets
+
+Gadgets offer the ability to save their current values as a user preset, or to load presets.
+
+```js
+{
+  "id": ___,
+  "userId": ___,
+  "name": "Pride Flag",
+  "type": "Palette",      // the type of the gadget
+  "data": { ... },        // type-dependent
+}
+```
+
+#### Color Gadget Data
+```js
+{
+  "color": 0xfedbca,
+}
+```
+
+#### Palette Gadget Data
+```js
+{
+  "type": "analogous" | "monochromatic" | "triad" | "complementary" | "compound" | "shades" | "custom",
+
+  // e.g. for type "custom":
+  "values": [0xE70000, 0xFF8C00, 0xFFEF00, 0x00811F, 0x0044FF, 0x760089],
+}
+
+```
+
+#### Slider Gadget Data
+```js
+{
+  "value": 0.7325,
+}
+```
+
+### Show Presets
+
+All of the current gadget settings for a show can be saved as a preset, which then shows up in the show selector.
+
+```js
+{
+  "userId": ___,
+  "name": "Awesome",
+  "showType": "TexturesAndStuff",
+  "gadgets": [
+    {
+      "id": "texture",    // the show's id for the gadget
+      "type": "Texture",  // the type of the gadget
+      "data": { "file": "/textures/tiger-print.png" }, // type-dependent
+    }
+  ]
+}
+```
+
+### Moving Head Positions
+
+These represent a focal point for the moving heads.
+
+Currently I'm thinking the eyes work like this: in the UI, you can specify a base position for moving heads. When
+`show.nextFrame()` is called, each moving head is already set to that base position, and it can then be altered
+arbitrarily by the show.
+
+These are saved position presets. 
+
+```js
+{
+  "name": "Headlight Mode",
+  "target": { "x": ___, "y": ___, "z": ___ },
```

Comment:
I don't see why 3D coordinates would be useful for the eyes?

Why not just use x/y as pan/tilt for the eyes?

I feel like z could specify another thing, such as the radius of the light? Are we making a super extremely narrow spotlight, or perhaps want to set the light as a big flood light?

Do our lights even provide controls over the radius of light?

Reply:
You need 3d coordinates to translate to polar coordinates, cuz the point could be anywhere in 3d space around the head, including behind it. (BAAAHS' eyes only point ahead, obvs, but other moving heads wouldn't necessarily have that restriction.

Reply:
To your second point on controlling other moving head parameters: good point.

I guess I've been thinking of their focal point as the main configurable, but it's reasonable that you might want to configure other things if the show ignores them.

My general idea here is to give an initial degree of user control to his eye positioning, which shows can then modify in show-appropriate ways, or ignore entirely. I'm assuming lots of shows won't do anything too interesting with the eyes; some shows will do some delta effect like make his eyes wobble a bit, or move in shape around some focal point; some shows will want to take control of the entirely.

Maybe this is overcomplicating things and there should just be another gadget used to provide a position in 3d space, and shows can just hook that up to the eyes. I just suspect lots of shows won't get around to doing it.